monitor/internal/pod/monitor.go

```diff
@@ -196,47 +171,101 @@ func (m *PodMonitor) Resync(ctx context.Context) error {
 
-	// starting the manager is a bit awkward:
-	// - it does not use contexts
-	// - we pass in a fake signal handler channel
-	// - we start another go routine which waits for the context to be cancelled
-	//   and closes that channel if that is the case
 	startTimestamp := time.Now()
 	z := make(chan struct{})
-	errCh := make(chan error, 2)
-	go func() {
-		<-ctx.Done()
-		close(z)
-		errCh <- ctx.Err()
-	}()
+	controllerStarted := make(chan struct{})
+
 	go func() {
-		if err := mgr.Start(z); err != nil {
-			errCh <- err
+		// manager.New already contacts the Kubernetes API
+		for {
+			var err error
+			mgr, err = managerNew(m.kubeCfg, manager.Options{})
+			if err != nil {
+				zap.L().Error("pod: new manager instantiation failed. Retrying in 3s...", zap.Error(err))
+				time.Sleep(retrySleep)
+				continue
+			}
+			break
+		}
+
+		// Create the delete event controller first
+		dc := NewDeleteController(mgr.GetClient(), m.localNode, m.handlers, m.sandboxExtractor, m.eventsCh)
+		for {
+			if err := mgr.Add(dc); err != nil {
+				zap.L().Error("pod: adding delete controller failed. Retrying in 3s...", zap.Error(err))
+				time.Sleep(retrySleep)
+				continue
+			}
+			break
+		}
+
+		// Create the main controller for the monitor
+		for {
+			if err := addController(
+				mgr,
+				newReconciler(mgr, m.handlers, m.metadataExtractor, m.netclsProgrammer, m.sandboxExtractor, m.localNode, m.enableHostPods, dc.GetDeleteCh(), dc.GetReconcileCh(), m.resyncInfo),
+				m.workers,
+				m.eventsCh,
+			); err != nil {
+				zap.L().Error("pod: adding main monitor controller failed. Retrying in 3s...", zap.Error(err))
+				time.Sleep(retrySleep)
+				continue
+			}
+			break
+		}
+
+		for {
+			if err := mgr.Add(&runnable{ch: controllerStarted}); err != nil {
+				zap.L().Error("pod: adding side controller failed. Retrying in 3s...", zap.Error(err))
+				time.Sleep(retrySleep)
+				continue
+			}
+			break
+		}
+
+		// starting the manager is a bit awkward:
+		// - it does not use contexts
+		// - we pass in a fake signal handler channel
+		// - we start another go routine which waits for the context to be cancelled
+		//   and closes that channel if that is the case
+
+		for {
+			if err := mgr.Start(z); err != nil {
+				zap.L().Error("pod: manager start failed. Retrying in 3s...", zap.Error(err))
+				time.Sleep(retrySleep)
+				continue
+			}
+			break
 		}
 	}()
 
 waitLoop:
 	for {
 		select {
-		case err := <-errCh:
-			return err
-		case <-time.After(5 * time.Second):
+		case <-ctx.Done():
+			close(z)
+		case <-time.After(warningMessageSleep):
 			// we give the controller 5 seconds to report back before we issue a warning
 			zap.L().Warn(startupWarningMessage)
 		case <-controllerStarted:
 			m.kubeClient = mgr.GetClient()
-			zap.L().Debug("pod: controller startup finished", zap.Duration("duration", time.Since(startTimestamp)))
+			zap.L().Error("pod: controller startup finished", zap.Duration("duration", time.Since(startTimestamp)))
```

Comment:
Why is this a Error? The controllers started successfully if it comes here. 
Am I missing something?

Reply:
ok, now I saw the runnable start function. Looks good, it should be Error. 
But I am wondering in what case can we come here. ???